library/bignum.c

```diff
@@ -499,26 +499,35 @@ int mbedtls_mpi_read_string( mbedtls_mpi *X, int radix, const char *s )
 }
 
 /*
- * Helper to write the digits high-order first
+ * Helper to write the digits high-order first.
  */
-static int mpi_write_hlp( mbedtls_mpi *X, int radix, char **p )
+static int mpi_write_hlp( mbedtls_mpi *X, int radix, char **p, const size_t buflen )
 {
     int ret;
     mbedtls_mpi_uint r;
+    size_t length = 0;
+    char *p_end = *p + buflen - 1;
 
     if( radix < 2 || radix > 16 )
         return( MBEDTLS_ERR_MPI_BAD_INPUT_DATA );
 
-    MBEDTLS_MPI_CHK( mbedtls_mpi_mod_int( &r, X, radix ) );
-    MBEDTLS_MPI_CHK( mbedtls_mpi_div_int( X, NULL, X, radix ) );
+    do
+    {
+        MBEDTLS_MPI_CHK( mbedtls_mpi_mod_int( &r, X, radix ) );
+        MBEDTLS_MPI_CHK( mbedtls_mpi_div_int( X, NULL, X, radix ) );
+        /*
+         * Write the residue in the current position, as an  ASCII character.
+         */
+        if( r < 10 )
+            *p_end-- = (char)( r + 0x30 );
+        else
+            *p_end-- = (char)( r + 0x37 );
 
-    if( mbedtls_mpi_cmp_int( X, 0 ) != 0 )
-        MBEDTLS_MPI_CHK( mpi_write_hlp( X, radix, p ) );
+        length++;
+    } while( mbedtls_mpi_cmp_int( X, 0 ) != 0 && length <= buflen );
```

Comment:
Blocker: If `length > buflen`, we'll leave the loop but nonetheless call `memmove()` with `length` as the length parameter.

We should fail if `length > buflen`.

Reply:
Same answer as https://github.com/ARMmbed/mbedtls/pull/2214#discussion_r234981909
I'll add a comment

Reply:
No, `length` is dynamic.

Reply:
But it starts from 0, and dependent on `buflen`.
`buflen` cannot be shorter than `n`, as checked in https://github.com/ARMmbed/mbedtls/blob/development/library/bignum.c#L552
`n` should be at most `length`

Reply:
In that case, one shouldn't check for `length <= buflen`. It is highly unintuitive and error-prone to have a buffer length parameter and nonetheless omit bounds checks because one _assumes_ it's large enough. There is technical justification here because of the `memmove()`, but nonetheless it's dangerous.

Please remove the check `length <= buflen` from the loop condition and instead check for `length < buflen` at the beginning of the loop body. This also removes the need for the precondition `buflen > 0`.

Reply:
I thought about not adding it at all, but then again, it's always better not to increase the size.
I'll make the change

Reply:
Yes, very true - if we can, we should be keep the code size down, and if it wasn't for the `memmove()` I'd be fine with using the fact that we have sufficient space as a precondition. But with the buffer length parameter in place, I think it's misleading to make this assumption.